tests/nn/data_parallel/test_fsdp_overlap.py

```diff
@@ -0,0 +1,207 @@
+            model.zero_grad(set_to_none=True)
+
+            cpu_iter_time = time.process_time() - cpu_start
+
+            # wait for gpu
+            out.item()
+            cpu_wait_for_gpu_time = time.process_time() - cpu_start - cpu_iter_time
+
+            # get sum of compute time
+            times = []
+            for mod in model.modules():
+                if not isinstance(mod, Layer):
+                    continue
+                times.append(mod.get_time())
+
+            # get gpu compute + all_gather time
+            overall_gpu_time = e1.elapsed_time(e2)
+
+            # print(f"rank {rank}", cpu_iter_time, cpu_wait_for_gpu_time, sum(times), "vs", e1.elapsed_time(e2))
+            cpu_iter.add(cpu_iter_time)
+            cpu_wait.add(cpu_wait_for_gpu_time)
+            gpu_compute.add(sum(times))
+            gpu_total.add(e1.elapsed_time(e2))
+
+        return [cpu_iter.avg(), cpu_wait.avg(), gpu_compute.avg(), gpu_total.avg()]
+
+    compute_cycles = 100_000_000
+    data_mb = 10
+    if fsdp_config["mixed_precision"]:
+        # make sure all-gather amount are the same in both mixed and full.
+        data_mb *= 2
+
+    e1 = run(0, 0)  # no compute, no all-gather
+    e2 = run(0, data_mb)  # no compute, only all-gather
+    e3 = run(compute_cycles, 0)  # only compute, no all-gather
+    e4 = run(compute_cycles, data_mb)  # both compute and all-gather
+    print(f"rank{rank}:\n  {e1}\n  {e2}\n  {e3}\n  {e4}")
+
+    # Check the cpu/gpu timing. CPU should run ahead of GPU. Therefore, cpu-gpu
+    # wait should be long, except when there is no real work on GPU.
+    #
+    # If the assertions fail below, we likely have a cpu-gpu wait in the forward/backward pass.
+    short = [e1[0], e2[0], e3[0], e4[0], e1[1]]
+    long = [e3[1], e4[1]]
+    if world_size == 1:
+        short.append(e2[1])  # all gather should not be happening.
+    else:
+        long.append(e2[1])  # all gather should happen and prolong the cpu-gpu wait.
+    for s in short:
+        for l in long:
+            # 10X longer is a safe margin, since the GPU work timing is around 100X more
+            # of that of the CPU.
+            assert s * 10 < l, f"{s} and {l}"
+
+    # Check the GPU timing.
+    short = [e1[2], e1[3], e2[2]]
+    long = [e3[2], e3[3], e4[2], e4[3]]
+    if world_size == 1:
+        short.append(e2[3])  # all gather should not be happening.
+    else:
+        long.append(e2[3])  # all gather should happen and prolong the cpu-gpu wait.
+    for s in short:
+        for l in long:
+            # 10X longer is a safe margin, since the time is around 100X longer
+            # when there is work on GPU vs. no work.
+            assert s * 10 < l, f"{s} and {l}"
+
+    # Check the GPU overlapping when there is all-gather.
+    if world_size > 1:
+        compute_only = e3[2]
+        all_gather_only = e2[3]
+        both = e4[3]
+        assert compute_only + all_gather_only > 1.1 * both, f"{compute_only} {all_gather_only} > 1.1 {both}"
```

Comment:
@myleott @QuentinDuval, please let me know if you have better ideas for unit testing this kind of issues.